File: adc01.ml
#require "Brulib"

open Brulib
module Stream = Brustream

(* let file = "test01.txt" *)
let file = "input01.txt"

let int_of_string_opt = function
  | "" -> None
  | s -> Some (int_of_string s)
;;

let _ =
  let data =
    Stream.(of_file file |> map int_of_string_opt)
    |> Da.of_stream
  in
  Da.push None data;
  let curr = ref 0
  and maxi = ref 0 in
  Da.iter
    (function
      | None ->
          maxi := max !maxi !curr;
          curr := 0
      | Some v -> curr := !curr + v)
    data;
  Printf.printf "Part One : %d\n" !maxi;
  (* Part Two *)
  let sums = Da.init ()
  and curr = ref 0 in
  Da.iter
    (function
      | None ->
          Da.heap_push (fun x y -> y - x) !curr sums;
          curr := 0
      | Some v -> curr := !curr + v)
    data;
  Printf.printf "Part Two : %d\n"
    (Da.heap_pop (fun x y -> y - x) sums
    + Da.heap_pop (fun x y -> y - x) sums
    + Da.heap_pop (fun x y -> y - x) sums)
;;

File: adc01.py
import heapq

data = []

with open("input01.txt", "r") as f:
    elf = 0
    for ligne in f:
        if ligne == "\n":
            data.append(-elf)
            elf = 0
        else:
            elf += int(ligne.strip())
    data.append(-elf)

heapq.heapify(data)
print("Part One:", maxi := - heapq.heappop(data))
print("Part Two:", maxi - heapq.heappop(data) - heapq.heappop(data))
